xhistogram/core.py

```diff
@@ -146,6 +147,64 @@ def _histogram_2d_vectorized(
     return bin_counts
 
 
+def _bincount(all_arrays, weights, axis, bins, density):
+
+    if weights is not None:
+        all_arrays.append(weights)
+
+    all_arrays_broadcast = broadcast_arrays(*all_arrays)
+
+    if weights is not None:
+        weights_broadcast = all_arrays_broadcast.pop()
+    else:
+        weights_broadcast = None
+    a0 = all_arrays_broadcast[0]
+
+    do_full_array = (axis is None) or (set(axis) == set(range(a0.ndim)))
+    if do_full_array:
+        kept_axes_shape = None
+    else:
+        kept_axes_shape = tuple([a0.shape[i] for i in range(a0.ndim) if i not in axis])
+
+    def reshape_input(a):
+        if do_full_array:
+            d = a.ravel()[None, :]
+        else:
+            # reshape the array to 2D
+            # axis 0: preserved axis after histogram
+            # axis 1: calculate histogram along this axis
+            new_pos = tuple(range(-len(axis), 0))
+            c = np.moveaxis(a, axis, new_pos)
+            split_idx = c.ndim - len(axis)
+            dims_0 = c.shape[:split_idx]
+            assert dims_0 == kept_axes_shape
+            dims_1 = c.shape[split_idx:]
+            new_dim_0 = np.prod(dims_0)
+            new_dim_1 = np.prod(dims_1)
+            d = reshape(c, (new_dim_0, new_dim_1))
+        return d
+
+    all_arrays_reshaped = [reshape_input(a) for a in all_arrays_broadcast]
+
+    if weights_broadcast is not None:
+        weights_reshaped = reshape_input(weights_broadcast)
+    else:
+        weights_reshaped = None
+
+    bin_counts = _bincount_2d_vectorized(
+        *all_arrays_reshaped, bins=bins, weights=weights_reshaped, density=density
+    )
```

Comment:
This is not obvious but: bin_counts is not necessarily 2D. Its ndim is 1 + len(all_arrays). So it represents and ND historgram.